neuralmonkey/encoders/sequence_cnn_encoder.py

```diff
@@ -0,0 +1,131 @@
+                 data_id: str,
+                 embedding_size: int,
+                 filters: List[Tuple[int, int]],
+                 max_input_len: Optional[int]=None,
+                 dropout_keep_prob: float=1.0,
+                 save_checkpoint: Optional[str]=None,
+                 load_checkpoint: Optional[str]=None) -> None:
+        """Creates a new instance of the CNN sequence encoder.
+
+        Based on: Yoon Kim: Convolutional Neural Networks for Sentence
+        Classification (http://emnlp2014.org/papers/pdf/EMNLP2014181.pdf)
+
+        Arguments:
+            vocabulary: Input vocabulary
+            data_id: Identifier of the data series fed to this encoder
+            name: An unique identifier for this encoder
+            max_input_len: Maximum length of an encoded sequence
+            embedding_size: The size of the embedding vector assigned
+                to each word
+            filters: Specification of CNN filters. It is a list of tuples
+                specifying the filter size and number of channels.
+            dropout_keep_prob: The dropout keep probability
+                (default 1.0)
+        """
+        ModelPart.__init__(self, name, save_checkpoint, load_checkpoint)
+
+        assert check_argument_types()
+
+        self.vocabulary = vocabulary
+        self.data_id = data_id
+        self.max_input_len = max_input_len
+
+        with tf.variable_scope(self.name):
+            self.train_mode = tf.placeholder(tf.bool, shape=[],
+                                             name="mode_placeholder")
+
+            self.inputs = tf.placeholder(tf.int32,
+                                         shape=[None, None],
+                                         name="encoder_input")
+
+            self._input_mask = tf.placeholder(
+                tf.float32, shape=[None, None],
+                name="encoder_padding")
+
+            with tf.variable_scope('input_projection'):
+                self.embedding_matrix = tf.get_variable(
+                    "word_embeddings", [len(vocabulary), embedding_size],
+                    initializer=tf.random_normal_initializer(stddev=0.01))
+                embedded_inputs = dropout(
+                    tf.nn.embedding_lookup(self.embedding_matrix, self.inputs),
+                    dropout_keep_prob,
+                    self.train_mode)
+
+            pooled_outputs = []
+            for filter_size, num_filters in filters:
+                with tf.variable_scope("conv-maxpool-%s" % filter_size):
+                    # Convolution Layer
+                    filter_shape = [filter_size, embedding_size, num_filters]
+                    w_filter = tf.get_variable(
+                        "conv_W", filter_shape,
+                        initializer=tf.random_uniform_initializer(-0.5, 0.5))
+                    b_filter = tf.get_variable(
+                        "conv_bias", [num_filters],
+                        initializer=tf.constant_initializer(0.0))
```

Comment:
nevim, ale mozna zeros initializer bude rychlejsi. ale je to asi ten debilni druh optimalizace